src/Microsoft.Azure.SignalR.Common/ClientInvocation/ClientResultsManager.cs

```diff
@@ -0,0 +1,312 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#if NET7_0_OR_GREATER
+using System.Threading;
+using System.Diagnostics;
+using System.Collections.Generic;
+using System.Collections.Concurrent;
+using System.Threading.Tasks;
+using System.Buffers;
+using System;
+using Microsoft.Azure.SignalR.Protocol;
+using Microsoft.AspNetCore.SignalR.Protocol;
+using Microsoft.AspNetCore.SignalR;
+
+namespace Microsoft.Azure.SignalR
+{
+    internal class ClientResultsManager : IClientResultsManager, IInvocationBinder
+    {
+        private readonly ConcurrentDictionary<string, (Type Type, string ConnectionId, object Tcs, Action<object, CompletionMessage> Complete)> _pendingInvocations = new();
+        private ulong _lastInvocationId = new();
+
+        private readonly ConcurrentDictionary<string, (Type Type, string ConnectionId, string CallerServerId, object Tcs, Action<object, CompletionMessage> Complete)> _routedInvocations = new();
+
+        private readonly ConcurrentDictionary<string, ServiceMappingMessage> _serviceMappingMessages = new();
+        private readonly IHubProtocolResolver _hubProtocolResolver;
+
+        public ClientResultsManager(IHubProtocolResolver hubProtocolResolver)
+        {
+            _hubProtocolResolver = hubProtocolResolver;
+        }
+
+        public ulong GetNewInvocation()
+        {
+            return Interlocked.Increment(ref _lastInvocationId);
+        }
+
+        public void AddServiceMappingMessage(string invocationId, ServiceMappingMessage serviceMappingMessage)
+        {
+            _serviceMappingMessages.TryAdd(invocationId, serviceMappingMessage);
+        }
+
+        public void RemoveServiceMappingMessageWithOfflinePing(string instanceId)
+        {
+            foreach (var (_, serviceMappingMessage) in _serviceMappingMessages)
+            {
+                if (serviceMappingMessage.InstanceId == instanceId)
```

Comment:
is not equal possible, shall we throw if not expected?

Reply:
`serviceMappingMessage.InstanceId != instanceId` is a possible and normal case. Because a server can behave as a route server for many different servers. So throwing exception is not needed.